src/multiplicity/mod.rs

```diff
@@ -0,0 +1,824 @@
+            .with(reduce.clone())
+            .input(input.0.clone())
+            .build();
+
+        // compute the counts of line weight
+        let counts_line_weight = &format!("_counts{LINE_WEIGHT}")[..];
+        let group_by_line_weight: Relation = Relation::reduce()
+            .with((LINE_WEIGHT, Expr::first(Expr::col(LINE_WEIGHT))))
+            .with((counts_line_weight, Expr::count(Expr::col(LINE_WEIGHT))))
+            .group_by(Expr::col(LINE_WEIGHT))
+            .input(new_reduce.clone())
+            .build();
+
+        let one_where_f1 = Expr::case(
+            Expr::eq(Expr::col(LINE_WEIGHT), Expr::val(1)),
+            Expr::val(1),
+            Expr::val(0),
+        );
+        let one_where_fj = Expr::case(
+            Expr::gt(Expr::col(LINE_WEIGHT), Expr::val(1)),
+            Expr::col(counts_line_weight),
+            Expr::val(0),
+        );
+
+        let tmp_map: Relation = Relation::map()
+            .with((LINE_WEIGHT, Expr::col(LINE_WEIGHT)))
+            .with(("_tmp_f1", one_where_f1))
+            .with(("_tmp_fj", one_where_fj))
+            .input(group_by_line_weight)
+            .build();
+
+        // compute f1: number of distinct values which occur exactly 1 times
+        // sum_fj: sum over number of distinct values which occur exactly j times
+        let tmp_reduce: Relation = Relation::reduce()
+            .with(("f1", Expr::sum(Expr::col("_tmp_f1"))))
+            .with(("sum_fj", Expr::sum(Expr::col("_tmp_fj"))))
+            .input(tmp_map)
+            .build();
+
+        let estimated_groups =
+            Expr::multiply(Expr::sqrt(Expr::val(input.1)), Expr::col("f1")) + Expr::col("sum_fj");
+
+        let estimated_and_actual_groups: Relation = Relation::map()
+            .with(("e", estimated_groups))
+            .with(("_actual_groups_", Expr::col("f1") + Expr::col("sum_fj")))
+            .input(tmp_reduce)
+            .build();
+
+        let correction_factor = Expr::divide(Expr::col("e"), Expr::col("_actual_groups_"));
+        let correction_factor_map: Relation = Relation::map()
+            .with(("_count_correction_factor_", correction_factor))
+            .filter(Expr::and(
+                Expr::gt_eq(Expr::col("e"), Expr::val(1.0)),
+                Expr::gt_eq(Expr::col("_actual_groups_"), Expr::val(1.0)),
+            ))
+            .limit(1)
+            .input(estimated_and_actual_groups)
+            .build();
+
+        // join correction_factor_map with the new reduce.
+        // correction_factor_map has 1 line result so the join type would be cross
+        let reduce_with_correction: Relation = Relation::join()
+            .left_names(
+                new_reduce
+                    .schema()
+                    .iter()
+                    .map(|f| f.name().to_string())
+                    .collect(),
+            )
+            .right_names(vec!["_count_correction_factor_".to_string()])
```

Comment:
What justifies to use lowercase here where we seem to use _CAPS_AND_UNDERSCORES_ style for reserved names in the relation?

Can you make the convention clearer / simpler?

Reply:
I used constants in CAPS_AND_UNDERSCORES style for all the intermediate relations. I also changed few names to be more meaningful.